### PR TITLE
L6 concurrent processing

### PR DESCRIPTION
**Pull request by DaniilShkurski**

### CODE REVIEW

File: modules/multiprocessingview/src/main/scala/com/itechart/concurrent/TripStatsConcurrentView.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  val logger: Logger = LogManager.getLogger(getClass)
+
+  def calculate(filedRecords: Seq[Future[Seq[Record]]]): Map[MonthKey, Double] = {
+    logger.info("concurrent GeneralStats started")
+
+    val res = filedRecords.map{fileRecord =>
+      fileRecord.map({records =>
+        val filteredRecords = records.filter(isValidRecord).map(record => TripStatsRecord(record.startDate, record.endDate, record.passengerCount))
+        val usageStats = TripStats.runConcurrent(filteredRecords)
+        usageStats
```
Comment (sleshJdev) [inline modules/multiprocessingview/src/main/scala/com/itechart/concurrent/TripStatsConcurrentView.scala]: `usageStats` is redundant here

File: modules/core/src/main/scala/com/itechart/core/GeneralStats.scala [hunk 2]
```diff
@@ -9,46 +9,75 @@ import org.apache.logging.log4j.{LogManager, Logger}
...
 
   final case class GeneralStatsRecord(startDateTime: LocalDateTime, endDateTime: LocalDateTime, tripDistance: Float, totalAmount: Float)
-  final case class GeneralStatsResult(rideMidSum: Double, longestByTimeRideDistance: Float, maximumRideAmountInterval:(LocalDateTime, LocalDateTime), recordAmount: Int, nonValidRecordAmount: Int)
+  final case class GeneralStatsResult(
+                                       rideMidSum: Double,
+                                       longestByTimeRideDistance: Float,
+                                       maximumRideAmountInterval:(LocalDateTime, LocalDateTime),
+                                       recordAmount: Int,
+                                       nonValidRecordAmount: Int)
+  final case class GeneralStatsResultConcurrent(
```
[the remark this answers is not in the document]
Comment (DaniilShkurski) [reply modules/core/src/main/scala/com/itechart/core/GeneralStats.scala]: Decided to rename it to GeneralStatsResultChunk:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/core/src/main/scala/com/itechart/core/GeneralStats.scala#L18-L26

File: modules/core/src/main/scala/com/itechart/core/TripStats.scala [hunk 3]
```diff
@@ -17,6 +17,11 @@ object TripStats {
   val logger: Logger = LogManager.getLogger(getClass)
 
   def run(tripStatsRecords: Seq[TripStatsRecord]): Map[MonthKey, Double] = {
+   val resultConcurrent = runConcurrent(tripStatsRecords)
+    resultConcurrent.view.mapValues(elem => elem._1.toDouble / elem._2).toMap
```
[the remark this answers is not in the document]
Comment (DaniilShkurski) [reply modules/core/src/main/scala/com/itechart/core/TripStats.scala]: Oops, forgot to fix it before, now fixed:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/core/src/main/scala/com/itechart/core/TripStats.scala#L21-L23

File: modules/core/src/main/scala/com/itechart/core/TripStats.scala [hunk 4]
```diff
@@ -17,6 +17,11 @@ object TripStats {
...
+  }
+
+  def runConcurrent(tripStatsRecords: Seq[TripStatsRecord]): Map[MonthKey, (Long, Int)] = {
```
[the remark this answers is not in the document]
Comment (DaniilShkurski) [reply modules/core/src/main/scala/com/itechart/core/TripStats.scala]: Renamed:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/core/src/main/scala/com/itechart/core/TripStats.scala#L26

File: modules/multiprocessingview/src/main/scala/com/itechart/concurrent/TripStatsConcurrentView.scala [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      })
+    }
+    val nonCombinedResult = Await.result(Future.sequence(res), Duration.Inf)
```
[the remark this answers is not in the document]
Comment (DaniilShkurski) [reply modules/multiprocessingview/src/main/scala/com/itechart/concurrent/TripStatsConcurrentView.scala]: Changed to returning Future value:
https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/multiprocessingview/src/main/scala/com/itechart/concurrent/TripStatsConcurrentView.scala#L15-L28

And waiting in main:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/multiprocessingview/src/main/scala/com/itechart/view/Main.scala#L58-L62

File: modules/multiprocessingview/src/main/scala/com/itechart/concurrent/GeneralStatsConcurrentView.scala [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+  def calculate(filedRecords: Seq[Future[Seq[Record]]]): Option[GeneralStatsResult] = {
+    logger.info("concurrent GeneralStats started")
+
+    val res = filedRecords.map{_.map(getGeneralStats)}
+    val futureRes = Future.sequence(res)
+
+    logger.info("Converted to sequence")
+
+    val nonCombinedResult = Await.result(futureRes, Duration.Inf)
```
[the remark this answers is not in the document]
Comment (DaniilShkurski) [reply modules/multiprocessingview/src/main/scala/com/itechart/concurrent/GeneralStatsConcurrentView.scala]: Fixed as all issues about waiting:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/multiprocessingview/src/main/scala/com/itechart/concurrent/GeneralStatsConcurrentView.scala#L29-L43

And:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/multiprocessingview/src/main/scala/com/itechart/view/Main.scala#L58-L62

File: modules/multiprocessingview/src/main/scala/com/itechart/concurrent/MissingDatesConcurrentView.scala [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+    logger.info("concurrent MissingDates started")
+
+    val completedFutures = filedRecords.map(_.map(
+      _.filter({!isValidRecord(_)})
+        .map(record => DateTimeRecord(record.startDate, record.endDate))
+    ))
+
+    logger.info("concurrent MissingDates completed")
+    Await.result(Future.sequence(completedFutures), Duration.Inf).flatten
```
[the remark this answers is not in the document]
Comment (DaniilShkurski) [reply modules/multiprocessingview/src/main/scala/com/itechart/concurrent/MissingDatesConcurrentView.scala]: Fixed as previous issues about returning futures:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/multiprocessingview/src/main/scala/com/itechart/concurrent/MissingDatesConcurrentView.scala#L15-L27

And:
https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/multiprocessingview/src/main/scala/com/itechart/view/Main.scala#L64-L67

File: modules/multiprocessingview/src/main/scala/com/itechart/concurrent/MissingDatesConcurrentView.scala [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+object MissingDatesConcurrentView {
+
+  val logger: Logger = LogManager.getLogger(getClass)
+
+  def calculate(filedRecords: Seq[Future[Seq[Record]]]): Seq[DateTimeRecord] = {
+
+    logger.info("concurrent MissingDates started")
+
+    val completedFutures = filedRecords.map(_.map(
+      _.filter({!isValidRecord(_)})
```
[the remark this answers is not in the document]
Comment (DaniilShkurski) [reply modules/multiprocessingview/src/main/scala/com/itechart/concurrent/MissingDatesConcurrentView.scala]: Fixed with using it:
https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/multiprocessingview/src/main/scala/com/itechart/concurrent/MissingDatesConcurrentView.scala#L20

File: modules/multiprocessingview/src/main/scala/com/itechart/concurrent/UsageStatsConcurrentView.scala [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+object UsageStatsConcurrentView {
+  val logger: Logger = LogManager.getLogger(getClass)
+
+  def calculate(filedRecords: Seq[Future[Seq[Record]]]): Map[LocalDate, Int] = {
+
+    logger.info("concurrent UsageStats started")
+
+    val res = filedRecords.map{fileRecord =>
+      fileRecord.map({records =>
+        val filteredRecords = records.filter(elem => isValidRecord(elem)).map(record => DateTimeRecord(record.startDate, record.endDate))
```
[the remark this answers is not in the document]
Comment (DaniilShkurski) [reply modules/multiprocessingview/src/main/scala/com/itechart/concurrent/UsageStatsConcurrentView.scala]: Also forgot to remove this before:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/multiprocessingview/src/main/scala/com/itechart/concurrent/UsageStatsConcurrentView.scala#L22

File: modules/multiprocessingview/src/main/scala/com/itechart/concurrent/TripStatsConcurrentView.scala (hunk 1 above)
Comment (DaniilShkurski) [reply modules/multiprocessingview/src/main/scala/com/itechart/concurrent/TripStatsConcurrentView.scala]: Removed usageStats:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/ad78c61297c05c5340cff732d0e368ad4bc5d296/modules/multiprocessingview/src/main/scala/com/itechart/concurrent/TripStatsConcurrentView.scala#L21

File: modules/core/src/main/scala/com/itechart/core/GeneralStats.scala [hunk 10]
```diff
@@ -9,46 +9,75 @@ import org.apache.logging.log4j.{LogManager, Logger}
...
+                                                 nonValidRecordAmount: Int)
+
   val logger: Logger = LogManager.getLogger(getClass)
   val intervalMinutesAmount: Int = 10
   val intervalSecondsAmount: Long = intervalMinutesAmount * TimeUnit.MINUTES.toSeconds(1)
 
-  def run(generalStatsRecords: Seq[GeneralStatsRecord], inputRecordsAmount: Int): Option[GeneralStatsResult] = {
+  def run(generalStatsRecords: Seq[GeneralStatsRecord], inputRecordsAmount: Int, firstIntervalStart: LocalDateTime): Option[GeneralStatsResult] = {
+    val generalStatsResultChunk = countSafeGeneralStatsChunk(generalStatsRecords, inputRecordsAmount, firstIntervalStart)
+    generalStatsResultChunk match{
```
Comment (sleshJdev) [inline modules/core/src/main/scala/com/itechart/core/GeneralStats.scala]: formatting

File: modules/core/src/main/scala/com/itechart/core/TripStats.scala [hunk 11]
```diff
@@ -28,13 +35,10 @@ object TripStats {
...
 
-    tripDurationByMonth.view.mapValues({
-      case (durationSum, amount) =>
-        durationSum.toDouble / amount
-    }).toMap
+    tripDurationByMonth
   }
 
-  def getStatsByMonth(record: TripStatsRecord): Seq[TripMonthStatsRecord] = {
+  def divideStatsByMonth(record: TripStatsRecord): Seq[TripMonthStatsRecord] = {
```
Comment (sleshJdev) [inline modules/core/src/main/scala/com/itechart/core/TripStats.scala]: Yeah, good naming

File: modules/core/src/main/scala/com/itechart/core/TripStats.scala [hunk 12]
```diff
@@ -17,9 +17,16 @@ object TripStats {
   val logger: Logger = LogManager.getLogger(getClass)
 
   def run(tripStatsRecords: Seq[TripStatsRecord]): Map[MonthKey, Double] = {
+   val resultConcurrent = computeTripMonthStats(tripStatsRecords)
```
Comment (sleshJdev) [inline modules/core/src/main/scala/com/itechart/core/TripStats.scala]: Why it's called resultConcurrent? Do you really need this val?

File: modules/multiprocessingview/src/main/scala/com/itechart/concurrent/GeneralStatsConcurrentView.scala [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+    futureRes.map(safeCombine)
+  }
+
+  def getGeneralStats(records: Seq[Record]): Option[GeneralStatsResultChunk] = {
+
+    logger.info("filtering started")
+
+    val filteredRecords = records.filter(isValidRecord)
+    val resFiltering = filteredRecords.map(record => GeneralStatsRecord(record.startDate, record.endDate, record.tripDistance, record.totalAmount))
```
Comment (sleshJdev) [inline modules/multiprocessingview/src/main/scala/com/itechart/concurrent/GeneralStatsConcurrentView.scala]: Can we merge it into one pipeline? Those intermediate vals don't bring any utility.

File: modules/core/src/main/scala/com/itechart/core/GeneralStats.scala (hunk 10 above)
Comment (DaniilShkurski) [reply modules/core/src/main/scala/com/itechart/core/GeneralStats.scala]: fixed:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/38e31d1c688562cc588410d81bab7925ea235007/modules/core/src/main/scala/com/itechart/core/GeneralStats.scala#L35

File: modules/core/src/main/scala/com/itechart/core/TripStats.scala (hunk 11 above)
Comment (DaniilShkurski) [reply modules/core/src/main/scala/com/itechart/core/TripStats.scala]: Changed name to better response its functionality:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/38e31d1c688562cc588410d81bab7925ea235007/modules/core/src/main/scala/com/itechart/core/TripStats.scala#L47

File: modules/core/src/main/scala/com/itechart/core/TripStats.scala (hunk 12 above)
Comment (DaniilShkurski) [reply modules/core/src/main/scala/com/itechart/core/TripStats.scala]: Removed this value:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/38e31d1c688562cc588410d81bab7925ea235007/modules/core/src/main/scala/com/itechart/core/TripStats.scala#L23-L30

File: modules/multiprocessingview/src/main/scala/com/itechart/concurrent/GeneralStatsConcurrentView.scala (hunk 13 above)
Comment (DaniilShkurski) [reply modules/multiprocessingview/src/main/scala/com/itechart/concurrent/GeneralStatsConcurrentView.scala]: Merged them:

https://github.com/DaniilShkurski/NYCGreenTaxi/blob/38e31d1c688562cc588410d81bab7925ea235007/modules/multiprocessingview/src/main/scala/com/itechart/concurrent/GeneralStatsConcurrentView.scala#L36-L39

**Pull request merged**